design/one-pager-stack-ui-metadata.md

```diff
@@ -0,0 +1,155 @@
+* **Stacks**: Stacks could refer to more than one Crossplane extension-manager packages or the Extension-Manager system as a whole.
+
+See the [Stacks Design Doc](design-doc-stacks.md) for more details about Crossplane Stacks.
+
+## Objective
+
+This document aims to provide details on the necessary metadata that will drive the UI on a configuration page with respect to Crossplane Stacks and the user configurable custom-resource fields exposed with a Stack. This is meant to be a dynamic spec that will not attempt to accomodate all potentional UI elements.  This spec does not intend to be as complete as [XForms](https://en.wikipedia.org/wiki/XForms) or [HTML Forms](https://www.w3.org/TR/html52/sec-forms.html), for example.
+
+Users may choose to annotate Stack bundled CRD today with UI hints.  This proposal suggests a means for the extension manager to apply CRD annotations at install and upgrade time.  This separation provides a better developer experience preventing the need for modifying escaped, nested, quoted, YAML or JSON annotations within a CRD document while iterating over UI design.
+
+A YAML file named `ui-schema.yaml` is proposed for inclusion in Stacks.  This will standardize the approach Stack creators use to offer UI context to their package and offer a more engaging user experience.
+
+Tools that deliver a user-interface based on this proposal may include:
+
+* CLI help output
+* CLI auto-completion for arguments
+* Advanced validation (beyond what Kubernetes' OAS3 support delivers)
+* ncurses-style installers
+* web-based package managers (out-of-cluster and in-cluster)
+* web-based configuration panels
+
+This spec is not concerned with a specific UI implementation but serves to provide a foundation for supporting Stack developers and users through basic form input.
+
+## Proposed File format
+
+The `ui-schema.yaml` file will conform to a specification that will not be covered formally in this document but will be suggested through examples and supporting text.
+
+## Proposed UI Elements
+
+* Sections
+* Title
+* Subtitle
+* Input: Text, Number, Password, Hidden, Readonly
+* Textarea
+* Select Dropdown: Single, Multiple
+* Checkbox: Single, Group
+* Radio Button Group
+
+## Stack Implementation
+
+The optional addition of a `ui-schema.yaml` file within the package tree will be used to drive the UI.
+
+The primary purpose of the `ui-schema.yaml` format is to allow for easy authoring of the definition of UI markup, validation, and errors for a more complete UI and UX. The `ui-schema.yaml` file will be parsed and validated as part of package validation and ultimately serialized as a `JSON` annotation on the respective CRD at install time.
+
+### Example addition of `ui-schema.yaml` in a package tree
+
+```
+.registry/
+├── icon.png
+├── app.yaml # Application metadata.
+├── ui-schema.yaml # Optional UI spec for configuration metadata
```

Comment:
what does a `ui-schema.yaml` file at the root mean vs. one that is scoped to a single CRD?  How would one at the root be processed and consumed?

Reply:
I've wondered if there is may be reasons to annotate the Extension itself in addition to the extension created CRD, for visual package-configuration aids (like debconf).

@rathpc, what did you have in mind here?

Reply:
This is a valid question. The inclusion of these different specs was originally added by @lukeweber and I believe the intention was to have a fallback during the consumption/overlay process that creates the annotation. If a specific CRD version does not contain the `ui-schema.yaml` file it would look for a root fallback to consume instead. This is really only necessary if a CRD has a mandatory requirement for a ui field, version agnostic. That is why it is also listed as optional.

Let me know if you need further clarification on this or if you guys have other thoughts. Thanks!

Reply:
I added this reasoning in 13b4da9 ..  I'm still wondering if Extensions themselves (and not the CRDs they provide/manage) may be able to take advantage of UI markup. (Configuring your Provider's credentials in an Infra Stack, for example).